File: Egg Dropping Puzzle - GFG/egg-dropping-puzzle.java
// { Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;

class GfG { 
    
	public static void main (String[] args) throws IOException  {
	    
	    //reading input using BufferedReader class
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		//reading total testcases
		int t = Integer.parseInt(br.readLine().trim());
		while(t-->0){
		    
		    //reading number of eggs and floors
		    String inputLine[] = br.readLine().trim().split(" ");
		    
		    int n = Integer.parseInt(inputLine[0]);
		    int k = Integer.parseInt(inputLine[1]);
		    
		    //calling eggDrop() method of class
		    //EggDrop
		    System.out.println(new Solution().eggDrop(n, k));
		}
	}
}


// } Driver Code Ends


class Solution 
{
    //Function to find minimum number of attempts needed in 
    //order to find the critical floor.
    static int eggDrop(int n, int k) 
	{
	    // Your code here
	    int[][] dp = new int[n + 1][k + 1];
	        
	    for(int i=0;i<=n;i++) {
	        for(int j=0;j<=k;j++) {
	            
	            //if floors or egg is zero return 0.
	            if(i == 0 || j == 0) 
	                dp[i][j] = 0;
	            
	            //if floor is 1 return 1.
	            else if(j == 1)
	                dp[i][j] = 1;
	                
	            //if egg is 1 we need to check at each floor.
	            else if(i == 1)
	                dp[i][j] = j;
	           
	            else {
	                int ans = Integer.MAX_VALUE;
	                for(int f=1;f<=j;f++) {
	                    int val = 1 + Math.max(dp[i - 1][f - 1], dp[i][j - f]);
	                    ans = Math.min(ans, val);
	                }
	                dp[i][j] = ans;
	            }
	        }
	    }
	    return dp[n][k];
	}
}

File: Egg Dropping Puzzle - GFG/egg-dropping-puzzle.cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
class Solution
{
    int dp[201][201];
    public:
    //Function to find minimum number of attempts needed in 
    //order to find the critical floor.
    int eggDrop(int n, int k) {
        memset(dp, -1, sizeof(dp));
        return EggDrop(n, k);
    }
    int EggDrop(int n, int k) 
    {
        if(k == 0 || k == 1) return k;
        if(n == 1) return k;
        
        if(dp[n][k] != -1) return dp[n][k];
        
        int minValue = INT_MAX, temp, i;
        
        for(i=1;i<=k;i++) {
            int lhs = dp[n - 1][i - 1] != -1 ? dp[n - 1][i - 1] : EggDrop(n - 1, i - 1);
            int rhs = dp[n][k - i] != -1 ? dp[n][k - i] : EggDrop(n, k - i);
            temp = 1 + max(lhs, rhs);
            minValue = min(minValue, temp);
        }
        
        return dp[n][k] = minValue;
    }
};

// { Driver Code Starts.
int main()
{
    //taking total testcases
    int t;
    cin>>t;
    while(t--)
    {
        //taking eggs and floors count
        int n, k;
        cin>>n>>k;
        Solution ob;
        //calling function eggDrop()
        cout<<ob.eggDrop(n, k)<<endl;
    }
    return 0;
}
  // } Driver Code Ends